### PR TITLE
feat: add pickle scanning docs

### PR DESCRIPTION
**Pull request by McPatate**
The docs should be updated again when we implement the features proposed in 'Potential Solutions' section.

### CODE REVIEW

File: docs/hub/security-pickle.md [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+
+To sum up, pickle is dangerous because :
+
+- when importing a python module, arbitrary code can be executed
+- you can import builtin functions like `eval` or `exec`, which can be used to execute arbitrary code
+- when instantiating an object, the constructor may be called
+
+This is why it is stated in most docs using pickle, do not unpickle data from untrusted sources.
+
+## Mitigation Strategies
```
Comment (julien-c) [inline docs/hub/security-pickle.md:165]: can we move the mitigation strategies and recommandations to the top? Feels like most people will want a TL;DR on this

File: docs/hub/security-pickle.md [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+
+Sound advice Luc, but pickle is used profusely and isn’t going anywhere soon : finding a new format everyone is happy with and initiating the change will take some time.
+
+So what can we do for now ?
+
+### Commit signing
+
+On the Hub, you have the ability to [sign your commits with a GPG key](./security-gpg). This does **not** guarantee that your file is not infected, but it does guarantee the origin of the file.
+
+If you know and trust user A and the commit that includes the file on the hub is signed by user A’s GPG key, it’s pretty safe to assume that you can trust the file.
```
Comment (julien-c) [inline docs/hub/security-pickle.md:177]: Add the option of loading from HDF5 (ie the TF weights) and show how to do it for a few examples of models?

File: docs/hub/security-pickle.md [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+
+This is why it is stated in most docs using pickle, do not unpickle data from untrusted sources.
+
+## Mitigation Strategies
+
+***Don’t use pickle***
+
+Sound advice Luc, but pickle is used profusely and isn’t going anywhere soon : finding a new format everyone is happy with and initiating the change will take some time.
+
+So what can we do for now ?
```
Comment (julien-c) [inline docs/hub/security-pickle.md:171]: you can move that further down IMO

File: docs/hub/security-pickle.md [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+
+### Commit signing
```
Comment (julien-c) [inline docs/hub/security-pickle.md:173]:
```suggestion
### Load files from users and organizations you trust (and use Commit signing)
```

File: docs/hub/security-pickle.md [hunk 5]
```diff
@@ -0,0 +1,226 @@
+# Pickle Scanning
+
+## What is a pickle ?
```
Comment (osanseviero) [inline docs/hub/security-pickle.md:3]: Maybe we could have a summary section before this

File: docs/hub/security-pickle.md [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+   57: \x94 MEMOIZE    (as 0)
+   58: .    STOP
+highest protocol among opcodes = 4
+```
+
+Don’t worry too much about the instructions for now, just know that the [pickletools](https://docs.python.org/3/library/pickletools.html) module is very useful for analyzing pickles. It allows you to read the instructions in the file ***without*** executing any code.
+
+Pickle is not simply a serialization protocol, it allows more flexibility by giving the ability to users to run python code at de-serialization time. Doesn’t sound good, does it ?
+
+## Why is it dangerous ?
```
Comment (osanseviero) [inline docs/hub/security-pickle.md:49]:
```suggestion
## Why is it dangerous?
```

File: docs/hub/security-pickle.md [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+    2: \x95 FRAME      48
+   11: \x8c SHORT_BINUNICODE 'data I want to share with a friend'
+   57: \x94 MEMOIZE    (as 0)
+   58: .    STOP
+highest protocol among opcodes = 4
+```
+
+Don’t worry too much about the instructions for now, just know that the [pickletools](https://docs.python.org/3/library/pickletools.html) module is very useful for analyzing pickles. It allows you to read the instructions in the file ***without*** executing any code.
+
+Pickle is not simply a serialization protocol, it allows more flexibility by giving the ability to users to run python code at de-serialization time. Doesn’t sound good, does it ?
```
Comment (osanseviero) [inline docs/hub/security-pickle.md:47]:
```suggestion
Pickle is not simply a serialization protocol, it allows more flexibility by giving the ability to users to run python code at de-serialization time. Doesn’t sound good, does it?
```

File: docs/hub/security-pickle.md [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+
+## Why is it dangerous ?
+
+As we’ve stated above, de-serializing pickle means that code can be executed. But this comes with certain limitations : you can only reference functions and classes from the top level module ; you cannot embed them in the pickle file itself.
```
Comment (osanseviero) [inline docs/hub/security-pickle.md:51]:
```suggestion
As we’ve stated above, de-serializing pickle means that code can be executed. But this comes with certain limitations: you can only reference functions and classes from the top level module; you cannot embed them in the pickle file itself.
```

File: docs/hub/security-pickle.md [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+# hexyl payload.pkl
+┌────────┬─────────────────────────┬─────────────────────────┬────────┬────────┐
+│00000000│ 80 04 95 33 00 00 00 00 ┊ 00 00 00 8c 08 5f 5f 6d │×•×30000┊000×•__m│
+│00000010│ 61 69 6e 5f 5f 94 8c 04 ┊ 44 61 74 61 94 93 94 29 │ain__××•┊Data×××)│
+│00000020│ 81 94 7d 94 8c 0f 69 6d ┊ 70 6f 72 74 61 6e 74 5f │××}××•im┊portant_│
+│00000030│ 73 74 75 66 66 94 8c 02 ┊ 34 32 94 73 62 2e       │stuff××•┊42×sb.  │
+└────────┴─────────────────────────┴─────────────────────────┴────────┴────────┘
+```
+
+We can see that there isn’t much in there, a few opcodes and the associated data. You might be thinking, so what’s the problem with pickle ?
```
Comment (osanseviero) [inline docs/hub/security-pickle.md:85]:
```suggestion
We can see that there isn’t much in there, a few opcodes and the associated data. You might be thinking, so what’s the problem with pickle?
```

File: docs/hub/security-pickle.md (hunk 3 above)
Comment (McPatate) [reply docs/hub/security-pickle.md:171]: I'd either leave it where it is or remove it entirely, what do you prefer ?

File: docs/hub/security-pickle.md [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+
+p.insert_python_exec('print("you\'ve been pwned !")')
+
+with open('payload.pkl', 'wb') as f:
+    p.dump(f)
+
+# innocently unpickle and get your friend's data
+with open('payload.pkl', 'rb') as f:
+    data = pickle.load(f)
+		print(data)
```
Comment (osanseviero) [inline docs/hub/security-pickle.md:108]:
```suggestion
with open('payload.pkl', 'rb') as f:
    data = pickle.load(f)
    print(data)
```

File: docs/hub/security-pickle.md (hunk 2 above)
Comment (osanseviero) [inline docs/hub/security-pickle.md:177]:
```suggestion
If you know and trust user A and the commit that includes the file on the Hub is signed by user A’s GPG key, it’s pretty safe to assume that you can trust the file.
```

File: docs/hub/security-pickle.md [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+
+On the Hub, you have the ability to [sign your commits with a GPG key](./security-gpg). This does **not** guarantee that your file is not infected, but it does guarantee the origin of the file.
+
+If you know and trust user A and the commit that includes the file on the hub is signed by user A’s GPG key, it’s pretty safe to assume that you can trust the file.
+
+### Hub’s Security Scanner
+
+#### What we have now
+
+We have created a security scanner that scans every file pushed to the hub and runs security checks. At the time of writing, it runs two types of scans :
```
Comment (osanseviero) [inline docs/hub/security-pickle.md:183]:
```suggestion
We have created a security scanner that scans every file pushed to the Hub and runs security checks. At the time of writing, it runs two types of scans :
```

**Review by osanseviero — COMMENTED** [review]
Thanks a lot for this! I left overall minor comments. I think some summary at the top would be super useful as the section on pickle security feels a bit as a tutorial and a big chunk of the audience might be more interested in the implications

File: docs/hub/security-pickle.md (hunk 6 above)
Comment (McPatate) [reply docs/hub/security-pickle.md:49]: The French in me, fixing it for all punctuation !

**Comment by McPatate** [comment]
> BTW also link to the Pytorch issue about safe loading of pickle, for reference?

[This](https://github.com/pytorch/pytorch/issues/52181) the one you're talking about ?

**Comment by adrinjalali** [comment]
We could also start experimenting with this: https://github.com/pytorch/torchsnapshot

**Review by adrinjalali — APPROVED** [review]
Nice addition. I don't have a strong opinion on the order of things here.

File: docs/hub/security-pickle.md (hunk 1 above)
Comment (McPatate) [reply docs/hub/security-pickle.md:181]: I've added a TOC and a summary at the top. Does this work better or do you want the mitigation strategies at the top ?

File: docs/hub/security-pickle.md (hunk 5 above)
Comment (McPatate) [reply docs/hub/security-pickle.md]: Is what I added ok or do you want something a bit lengthier ?

File: docs/hub/security-pickle.md (hunk 1 above)
Comment (julien-c) [reply docs/hub/security-pickle.md:181]: works for me to add this as a TL;DR like you did, but i would still allude to the main short-term mitigations

Comment (julien-c) [reply docs/hub/security-pickle.md:181]: (see my suggestion)

File: docs/hub/security-pickle.md [hunk 12]
```diff
@@ -0,0 +1,256 @@
...
+  * [Hub’s Security Scanner](#hubs-security-scanner)
+    + [What we have now](#what-we-have-now)
+    + [Potential solutions](#potential-solutions)
+- [Further Reading](#further-reading)
+
+## TL;DR
+
+Pickle is a widely used serialization format in ML. Most notably, it is the default format for PyTorch model weights.
+
+There are dangerous arbitrary code execution attacks that can be perpetrated when you load a pickle file.
```
Comment (julien-c) [inline docs/hub/security-pickle.md:18]:
```suggestion
There are dangerous arbitrary code execution attacks that can be perpetrated when you load a pickle file. We suggest loading models from users and organizations you trust, relying on signed commits, and/or loading models from TF or Jax formats with the `from_tf=True` auto-conversion mechanism. We will also alleviate this issue shortly by displaying/"vetting" the list of imports in any pickled file, directly on the Hub. 
```

Comment (julien-c) [reply docs/hub/security-pickle.md:18]: updating the TL;DR here, what do you think @McPatate @osanseviero @adrinjalali @rajshah4?

Comment (McPatate) [reply docs/hub/security-pickle.md:18]: lgtm !

Comment (julien-c) [reply docs/hub/security-pickle.md:18]: alright if everyone agrees let's merge and publish on twitter?

**Pull request merged**